### PR TITLE
INT-3741

### PR DESCRIPTION
**Pull request by fijijavis**
Changes in this PR:

1. Adding unit test coverage for `enum` and `model` areas of the application
2. Cleanup of how types for enums are being generated.  There was a material disconnect between the two that needed to be addressed.  Taking advantage of [Template Literal Type](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#template-literal-types) which were introduced in TS 4.1
3. Cleanup of `.npmignore` so we are only publishing the necessary files
4. Cleanup some naming inconsistencies with Job objects

### CODE REVIEW

File: .npmignore [hunk 1]
```diff
@@ -1,6 +1,12 @@
...
+test/
+.eslintignore
+.eslintrc.js
+.prettierrc.js
+commitlint.config.js
+jest-intg-config.ts
+jest.config.ts
+ts-auto-mock-config.ts
+tsconfig.json
+tsconfig.paths.json
```
Comment (fijijavis) [inline .npmignore:12]: none of these should be included with our published package

File: src/enum/data-set-type.ts [hunk 2]
```diff
@@ -11,28 +11,14 @@ export enum DataSetTypes {
   POWER_BI_DATAFLOW_ENTITY = 'POWER_BI_DATAFLOW_ENTITY',
   ADF_DATA_FLOW = 'ADF_DATA_FLOW',
   ADF_DATASET = 'ADF_DATASET',
-  FTP = 'FTP'
+  FTP = 'FTP',
+  UNKNOWN = 'UNKNOWN',
+  SYNAPSE_DATAFLOW = 'SYNAPSE_DATAFLOW',
+  PACKAGE = 'PACKAGE',
+  COMPONENT = 'COMPONENT',
+  DATASET = 'DATASET'
```
Comment (fijijavis) [inline src/enum/data-set-type.ts:19]: reconciling what was in the `DataSetType` but not in the enum

File: src/enum/data-set-type.ts [hunk 3]
```diff
@@ -11,28 +11,14 @@ export enum DataSetTypes {
...
-  | 'POWER_BI_DATASET'
-  | 'POWER_BI_DATAFLOW_ENTITY'
-  | 'ADF_DATA_FLOW'
-  | 'ADF_DATASET'
-  | 'SYNAPSE_DATAFLOW'
-  | 'PACKAGE'
-  | 'COMPONENT'
-  | 'DATASET'
-  | 'FTP';
+export type DataSetType = `${DataSetTypes}`;
```
Comment (fijijavis) [inline src/enum/data-set-type.ts:21]: generate `DataSetType` based on values in the enum so they are never out of sync going forward

File: src/enum/job-did-not-run-at-method.ts [hunk 4]
```diff
@@ -2,13 +2,12 @@ export enum JobDidNotRunAtMethods {
   TRIGGER_TYPE = 'TRIGGER_TYPE',
   NEXT_RUN_AT = 'NEXT_RUN_AT'
 }
+export type JobDidNotRunAtMethod = `${JobDidNotRunAtMethods}`;
```
Comment (fijijavis) [inline src/enum/job-did-not-run-at-method.ts:5]: generate `JobDidNotRunAtMethod` based on values in the enum so they are never out of sync going forward

File: src/enum/job-run-log-status.ts [hunk 5]
```diff
@@ -6,13 +6,12 @@ export enum JobRunLogStatuses {
   QUEUED = 'QUEUED',
   WARNING = 'WARNING'
 }
+export type JobRunLogStatus = `${JobRunLogStatuses}`;
```
Comment (fijijavis) [inline src/enum/job-run-log-status.ts:9]: generate `JobRunLogStatus` based on values in the enum so they are never out of sync going forward

File: src/enum/job-type.ts [hunk 6]
```diff
@@ -15,30 +15,17 @@ export enum JobTypes {
   COMPONENT = 'COMPONENT',
   TASK = 'TASK',
   SNOWFLAKE_PIPE = 'SNOWFLAKE_PIPE',
-  AWS_LAMBDA = 'AWS_LAMBDA'
+  AWS_LAMBDA = 'AWS_LAMBDA',
+  POWER_BI_ACTIVITY = 'POWER_BI_ACTIVITY',
+  DATAFLOW = 'DATAFLOW',
+  REFRESH = 'REFRESH'
```
Comment (fijijavis) [inline src/enum/job-type.ts:21]: reconciling what was in the type but not in the enum

File: src/enum/job-type.ts [hunk 7]
```diff
@@ -15,30 +15,17 @@ export enum JobTypes {
...
-  | 'JOB'
-  | 'DATAFLOW'
-  | 'PIPELINE'
-  | 'PACKAGE'
-  | 'COMPONENT'
-  | 'REFRESH'
-  | 'TASK'
-  | 'SNOWFLAKE_PIPE'
-  | 'AWS_LAMBDA';
+export type JobType = `${JobTypes}`;
```
Comment (fijijavis) [inline src/enum/job-type.ts:24]: generate `JobType` based on values in the enum so they are never out of sync going forward

File: src/enum/object-type.ts [hunk 8]
```diff
@@ -38,44 +38,7 @@ export enum ObjectTypes {
...
-  | 'DATASOURCE'
-  | 'DATASET'
-  | 'DATAFLOW'
-  | 'ACTIVITY'
-  | 'PIPELINE'
-  | 'REPORT'
-  | 'TRIGGER'
-  | 'REFRESH'
-  | 'TASK';
+export type ObjectType = `${ObjectTypes}`;
```
Comment (fijijavis) [inline src/enum/object-type.ts:41]: generate `ObjectType` based on values in the enum so they are never out of sync going forward

File: src/enum/object-type.ts [hunk 9]
```diff
@@ -107,26 +70,7 @@ export enum JobObjectTypes {
...
-  | 'ACTIVITY'
-  | 'JOB'
-  | 'DATAFLOW'
-  | 'PIPELINE'
-  | 'PACKAGE'
-  | 'COMPONENT'
-  | 'REFRESH'
-  | 'TASK'
-  | 'SNOWFLAKE_PIPE';
+export type JobObjectType = `${JobObjectTypes}`;
```
Comment (fijijavis) [inline src/enum/object-type.ts:73]: generate `JobObjectType` based on values in the enum so they are never out of sync going forward

File: src/enum/object-type.ts [hunk 10]
```diff
@@ -140,7 +84,7 @@ export enum DataVolumeObjectTypes {
   SQL_TABLE = 'SQL_TABLE'
 }
 
-export type DataVolumeObjectType = 'SQL_TABLE';
+export type DataVolumeObjectType = `${DataVolumeObjectTypes}`;
```
Comment (fijijavis) [inline src/enum/object-type.ts:87]: generate `DataVolumeObjectType` based on values in the enum so they are never out of sync going forward

File: src/enum/platform-type.ts [hunk 11]
```diff
@@ -28,30 +28,7 @@ export function getPlatformTypes(): string[] {
...
-  | 'INFORMATICA'
-  | 'PBIRS'
-  | 'EXTERNAL'
-  | 'AZSTORAGE'
-  | 'ORACLE'
-  | 'DATAPROC'
-  | 'BIGQUERY'
-  | 'COMPOSER'
-  | 'CUSTOM_LOGS';
+export type PlatformType = `${PlatformTypes}`;
```
Comment (fijijavis) [inline src/enum/platform-type.ts:31]: generate `PlatformType` based on values in the enum so they are never out of sync going forward

File: src/model/job/dbt-job.ts [hunk 12]
```diff
@@ -10,7 +10,7 @@ export type DBTJobConstructorParams = {
   name: string;
 };
 
-export class DBTJobConstructor extends Job {
+export class DBTJob extends Job {
```
Comment (fijijavis) [inline src/model/job/dbt-job.ts:13]: cleanup naming convention inconsistency

File: src/model/job/dbt-model.ts [hunk 13]
```diff
@@ -10,7 +10,7 @@ export type DBTModelConstructorParams = {
   jobName: string;
 };
 
-export class DBTModelConstructor extends Job {
+export class DBTModel extends Job {
```
Comment (fijijavis) [inline src/model/job/dbt-model.ts:13]: cleanup naming convention inconsistency

File: src/model/job/fivetran-connector.ts [hunk 14]
```diff
@@ -5,7 +5,7 @@ export type FivetranConnectorConstructorParams = {
   name: string;
 };
 
-export class FivetranConnectorConstructor extends Job {
+export class FivetranConnector extends Job {
```
Comment (fijijavis) [inline src/model/job/fivetran-connector.ts:8]: cleanup naming convention inconsistency

File: src/model/job/dbt-job.ts (hunk 12 above)
Comment (johnberberich) [reply src/model/job/dbt-job.ts:13]: Hypothetically, should we still support clients using `DBTJobConstructor`?

```suggestion
export class DBTJob extends Job {
  ...
 }
 
// Untested: 
const DBTJobConstructor = DBTJob;
export DBTJobConstructor;
```

Realistically, it doesn't look like our only client of this is using that class, so this probably doesn't matter. Also, a library upgrade wouldn't be forced upon them, so they could make this small change if they needed to upgrade, right?

File: package.json [hunk 15]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@pantomath/sdk-javascript",
-  "version": "1.2.3",
+  "version": "2.0.0",
```
Comment (johnberberich) [inline package.json:3]: Ahh, this would be a major version change. Ignore my comment about class renaming.

File: src/model/job/dbt-job.ts (hunk 12 above)
Comment (johnberberich) [reply src/model/job/dbt-job.ts:13]: Nevermind.

Comment (fijijavis) [reply src/model/job/dbt-job.ts:13]: Yeah, I think we make this a major version bump to be safe and in the release notes we can document the differences.

**Review by johnberberich — APPROVED** [review]
This is great, thanks Jim!

**Pull request merged**